File: ch02/src/cpp/compare_exchange_id_allocation.cpp
#include <atomic>
#include <cstddef>
#include <cstdint>
#include <iostream>
#include <limits>
#include <stdexcept>
#include <syncstream>
#include <thread>
#include <vector>

std::uint64_t allocate_new_id()
{
    static const auto kLimits { std::numeric_limits<std::uint64_t>::max() };

    static auto next_id = std::atomic_uint64_t { 0 };
    auto id = next_id.load(std::memory_order_relaxed);

    while (true) {
        // Check the limit before modifying next_id
        // to ensure that there is no overflow
        if (id == kLimits) {
            throw std::runtime_error { "Too many IDs!" };
        }

        if (next_id.compare_exchange_weak(id, id + 1, std::memory_order_relaxed)) {
            // OK, the value has been successfully changed
            return id + 1;
        } else {
            // The value has been changed by another thread, but not this one.
            // Try again in the next iteration.
        }
    }

    return id;
}

int main()
{
    static constexpr std::size_t kNumThreads { 20 };
    std::vector<std::jthread> handles;
    handles.reserve(kNumThreads);

    for (std::size_t i { 0 }; i < kNumThreads; ++i) {
        handles.emplace_back([]() {
            const auto id { allocate_new_id() };
            std::osyncstream { std::cout } << "Thread "
                                           << std::this_thread::get_id()
                                           << " -> "
                                           << id << '\n';
        });
    }

    return 0;
}

File: ch02/src/bin/compare_exchange_id_allocation.rs
use std::{
    sync::atomic::{AtomicU64, Ordering},
    thread,
};

fn main() {
    let mut handles = vec![];
    for _ in 0..20 {
        handles.push(thread::spawn(|| {
            let id = allocate_new_id();
            println!("Thread {:?} -> {}", thread::current().id(), id);
        }));
    }

    for handle in handles {
        handle.join().unwrap();
    }
}

fn allocate_new_id() -> u64 {
    // The maximum ID that should be exceeded
    const LIMIT: u64 = 1000;

    static NEXT_ID: AtomicU64 = AtomicU64::new(0);
    let mut id = NEXT_ID.load(Ordering::Relaxed);

    // Put in a loop to ensure the work is done,
    // when compare_exchange returns Ok(_).
    loop {
        // Check and panic before modifying NEXT_ID. This ensures that
        // NEXT_ID will never overflow the limit.
        assert!(id < LIMIT, "Too many IDs");

        match NEXT_ID.compare_exchange(id, id + 1, Ordering::Relaxed, Ordering::Relaxed) {
            Ok(_) => {
                // OK: NEXT_ID has not been modified by other threads
                // after we loaded it.
                return id;
            }
            Err(v) => {
                // NEXT_ID has been modified by some other thread(s)
                // after we loaded it. Try again with the updated value.
                id = v;
            }
        }
    }
}
